server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java

```diff
@@ -19,81 +19,342 @@
 
 package org.elasticsearch.search.internal;
 
+import org.elasticsearch.Version;
+import org.elasticsearch.action.IndicesRequest;
+import org.elasticsearch.action.OriginalIndices;
 import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchTask;
 import org.elasticsearch.action.search.SearchType;
+import org.elasticsearch.action.support.IndicesOptions;
 import org.elasticsearch.cluster.metadata.AliasMetaData;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.common.CheckedFunction;
 import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesArray;
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.index.Index;
 import org.elasticsearch.index.query.BoolQueryBuilder;
 import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
 import org.elasticsearch.index.query.Rewriteable;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.indices.AliasFilterParsingException;
 import org.elasticsearch.indices.InvalidAliasNameException;
 import org.elasticsearch.search.Scroll;
 import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.transport.TransportRequest;
 
 import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
 import java.util.function.Function;
 
 /**
  * Shard level request that represents a search.
- * It provides all the methods that the {@link org.elasticsearch.search.internal.SearchContext} needs.
+ * It provides all the methods that the {@link SearchContext} needs.
  * Provides a cache key based on its content that can be used to cache shard level response.
  */
-public interface ShardSearchRequest {
+public class ShardSearchRequest extends TransportRequest implements IndicesRequest {
+    private final String clusterAlias;
+    private final ShardId shardId;
+    private final int numberOfShards;
+    private final SearchType searchType;
+    private final Scroll scroll;
+    private final float indexBoost;
+    private final Boolean requestCache;
+    private final long nowInMillis;
+    private final boolean allowPartialSearchResults;
+    private final String[] indexRoutings;
+    private final String preference;
+    private final OriginalIndices originalIndices;
 
-    ShardId shardId();
+    //these are the only two mutable fields, as they are subject to rewriting
+    private AliasFilter aliasFilter;
+    private SearchSourceBuilder source;
 
-    SearchSourceBuilder source();
+    public ShardSearchRequest(SearchRequest searchRequest,
```

Comment:
Seems like only SliceBuilderTests uses this constructor. Can we change the caller to rely on the other constructor that accepts also the OriginalIndices argument?

Reply:
+1